Fails to bootstrap rust-1.54.0 on aarch64 / macOS 12

I'm running on a local root 2c4f89f87692f27246d1bc1e42c8d43df8514a71 `./build-1.54.0.sh` which is failed as:
```
> /Users/catap/src/mrustc/bin/mrustc -Z print-cfgs
(0/13) BUILDING core v0.0.0
> /Users/catap/src/mrustc/bin/mrustc rustc-1.54.0-src/library/core/src/lib.rs -o output-1.54.0/libcore.rlib --crate-name core --crate-type rlib -C emit-depfile=output-1.54.0/libcore.rlib.d --crate-tag 0_0_0 --cfg debug_assertions -O -L output-1.54.0 --edition 2018
:0: BUG:src/mir/from_hir.cpp:524: TODO: visit - asm! const
Process was terminated with signal 6
FAILING COMMAND:  /Users/catap/src/mrustc/bin/mrustc rustc-1.54.0-src/library/core/src/lib.rs -o output-1.54.0/libcore.rlib --crate-name core --crate-type rlib -C emit-depfile=output-1.54.0/libcore.rlib.d --crate-tag 0_0_0 --cfg debug_assertions -O -L output-1.54.0 --edition 2018
BUILD FAILED
make: *** [output-1.54.0/libstd.rlib] Error 1

```

@thepowersgang I'm using 11a9c01119988e30ba40bab4ee5f60753c0c4b61 as local root which had failed as:
```
(0/13) BUILDING core v0.0.0
> /Users/catap/src/mrustc/bin/mrustc rustc-1.54.0-src/library/core/src/lib.rs -o output-1.54.0/libcore.rlib --crate-name core --crate-type rlib -C emit-depfile=output-1.54.0/libcore.rlib.d --crate-tag 0_0_0 --cfg debug_assertions -O -L output-1.54.0 --edition 2018
output-1.54.0/libcore.rlib.c:113819:23: error: unknown directive
        __asm__ __volatile__(".intel_syntax; nop;\n .att_syntax; " : :);
                             ^
<inline asm>:1:2: note: instantiated into assembly here
        .intel_syntax; nop;
        ^
output-1.54.0/libcore.rlib.c:113819:45: error: unknown directive
        __asm__ __volatile__(".intel_syntax; nop;\n .att_syntax; " : :);
                                                   ^
<inline asm>:2:2: note: instantiated into assembly here
 .att_syntax; 
 ^
2 errors generated.
C Compiler failed to execute - error code 256
Process exited with non-zero exit status 1
```

Ok, it had failed on linking `cargo` as:
```
> /Users/catap/src/mrustc/bin/mrustc -Z print-cfgs
BUILDING cargo v0.55.0
> /Users/catap/src/mrustc/bin/mrustc rustc-1.54.0-src/src/tools/cargo/src/bin/cargo/main.rs -o output-1.54.0/cargo-build/cargo --crate-name cargo --crate-type bin -C emit-depfile=output-1.54.0/cargo-build/cargo.d --crate-tag 0_55_0 --cfg debug_assertions -O -L output-1.54.0 -L output-1.54.0/cargo-build --extern cargo=output-1.54.0/cargo-build/libcargo-0_55_0.rlib --edition 2018 --extern atty=output-1.54.0/cargo-build/libatty-0_2_14.rlib --extern bytesize=output-1.54.0/cargo-build/libbytesize-1_0_1.rlib --extern cargo_platform=output-1.54.0/cargo-build/libcargo_platform-0_1_1.rlib --extern cargo_util=output-1.54.0/cargo-build/libcargo_util-0_1_0.rlib --extern crates_io=output-1.54.0/cargo-build/libcrates_io-0_33_0.rlib --extern crossbeam_utils=output-1.54.0/cargo-build/libcrossbeam_utils-0_8_5_H7.rlib --extern curl=output-1.54.0/cargo-build/libcurl-0_4_36_H22.rlib --extern curl_sys=output-1.54.0/cargo-build/libcurl_sys-0_4_42_H46.rlib --extern env_logger=output-1.54.0/cargo-build/libenv_logger-0_8_3_Hf.rlib --extern anyhow=output-1.54.0/cargo-build/libanyhow-1_0_40_H2.rlib --extern filetime=output-1.54.0/cargo-build/libfiletime-0_2_14.rlib --extern flate2=output-1.54.0/cargo-build/libflate2-1_0_20_H411.rlib --extern git2=output-1.54.0/cargo-build/libgit2-0_13_17_H7.rlib --extern git2_curl=output-1.54.0/cargo-build/libgit2_curl-0_14_1.rlib --extern glob=output-1.54.0/cargo-build/libglob-0_3_0.rlib --extern hex=output-1.54.0/cargo-build/libhex-0_4_2_H2.rlib --extern home=output-1.54.0/cargo-build/libhome-0_5_3.rlib --extern humantime=output-1.54.0/cargo-build/libhumantime-2_0_1.rlib --extern ignore=output-1.54.0/cargo-build/libignore-0_4_17.rlib --extern lazy_static=output-1.54.0/cargo-build/liblazy_static-1_4_0.rlib --extern jobserver=output-1.54.0/cargo-build/libjobserver-0_1_22.rlib --extern lazycell=output-1.54.0/cargo-build/liblazycell-1_3_0.rlib --extern libc=output-1.54.0/cargo-build/liblibc-0_2_95_H21.rlib --extern log=output-1.54.0/cargo-build/liblog-0_4_14_H20000.rlib --extern libgit2_sys=output-1.54.0/cargo-build/liblibgit2_sys-0_12_18_H1f.rlib --extern memchr=output-1.54.0/cargo-build/libmemchr-2_4_0_H6.rlib --extern num_cpus=output-1.54.0/cargo-build/libnum_cpus-1_13_0.rlib --extern opener=output-1.54.0/cargo-build/libopener-0_4_1.rlib --extern percent_encoding=output-1.54.0/cargo-build/libpercent_encoding-2_1_0.rlib --extern rustfix=output-1.54.0/cargo-build/librustfix-0_5_1.rlib --extern semver=output-1.54.0/cargo-build/libsemver-1_0_3_H3.rlib --extern serde=output-1.54.0/cargo-build/libserde-1_0_126_H1a.rlib --extern serde_ignored=output-1.54.0/cargo-build/libserde_ignored-0_1_2.rlib --extern serde_json=output-1.54.0/cargo-build/libserde_json-1_0_64_He0.rlib --extern shell_escape=output-1.54.0/cargo-build/libshell_escape-0_1_5.rlib --extern strip_ansi_escapes=output-1.54.0/cargo-build/libstrip_ansi_escapes-0_1_0.rlib --extern tar=output-1.54.0/cargo-build/libtar-0_4_35.rlib --extern tempfile=output-1.54.0/cargo-build/libtempfile-3_2_0.rlib --extern termcolor=output-1.54.0/cargo-build/libtermcolor-1_1_2.rlib --extern toml=output-1.54.0/cargo-build/libtoml-0_5_7.rlib --extern unicode_xid=output-1.54.0/cargo-build/libunicode_xid-0_2_2.rlib --extern url=output-1.54.0/cargo-build/liburl-2_2_2_H1.rlib --extern walkdir=output-1.54.0/cargo-build/libwalkdir-2_3_2.rlib --extern clap=output-1.54.0/cargo-build/libclap-2_33_3_H260b.rlib --extern unicode_width=output-1.54.0/cargo-build/libunicode_width-0_1_8.rlib --extern im_rc=output-1.54.0/cargo-build/libim_rc-15_0_0.rlib --extern itertools=output-1.54.0/cargo-build/libitertools-0_10_0_H3.rlib --extern rustc_workspace_hack=output-1.54.0/cargo-build/librustc_workspace_hack-1_0_0.rlib --extern rand=output-1.54.0/cargo-build/librand-0_8_3_H1847.rlib
ld: reference to symbol (which has not been assigned an address) _ZRG2cE14curl0_4_36_H22B_09INIT_CTOR0g in '_ZRG1cE14curl0_4_36_H224init0g' from output-1.54.0/cargo-build/libcurl-0_4_36_H22.rlib.o for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
C Compiler failed to execute - error code 256
Process exited with non-zero exit status 1
FAILING COMMAND:  /Users/catap/src/mrustc/bin/mrustc rustc-1.54.0-src/src/tools/cargo/src/bin/cargo/main.rs -o output-1.54.0/cargo-build/cargo --crate-name cargo --crate-type bin -C emit-depfile=output-1.54.0/cargo-build/cargo.d --crate-tag 0_55_0 --cfg debug_assertions -O -L output-1.54.0 -L output-1.54.0/cargo-build --extern cargo=output-1.54.0/cargo-build/libcargo-0_55_0.rlib --edition 2018 --extern atty=output-1.54.0/cargo-build/libatty-0_2_14.rlib --extern bytesize=output-1.54.0/cargo-build/libbytesize-1_0_1.rlib --extern cargo_platform=output-1.54.0/cargo-build/libcargo_platform-0_1_1.rlib --extern cargo_util=output-1.54.0/cargo-build/libcargo_util-0_1_0.rlib --extern crates_io=output-1.54.0/cargo-build/libcrates_io-0_33_0.rlib --extern crossbeam_utils=output-1.54.0/cargo-build/libcrossbeam_utils-0_8_5_H7.rlib --extern curl=output-1.54.0/cargo-build/libcurl-0_4_36_H22.rlib --extern curl_sys=output-1.54.0/cargo-build/libcurl_sys-0_4_42_H46.rlib --extern env_logger=output-1.54.0/cargo-build/libenv_logger-0_8_3_Hf.rlib --extern anyhow=output-1.54.0/cargo-build/libanyhow-1_0_40_H2.rlib --extern filetime=output-1.54.0/cargo-build/libfiletime-0_2_14.rlib --extern flate2=output-1.54.0/cargo-build/libflate2-1_0_20_H411.rlib --extern git2=output-1.54.0/cargo-build/libgit2-0_13_17_H7.rlib --extern git2_curl=output-1.54.0/cargo-build/libgit2_curl-0_14_1.rlib --extern glob=output-1.54.0/cargo-build/libglob-0_3_0.rlib --extern hex=output-1.54.0/cargo-build/libhex-0_4_2_H2.rlib --extern home=output-1.54.0/cargo-build/libhome-0_5_3.rlib --extern humantime=output-1.54.0/cargo-build/libhumantime-2_0_1.rlib --extern ignore=output-1.54.0/cargo-build/libignore-0_4_17.rlib --extern lazy_static=output-1.54.0/cargo-build/liblazy_static-1_4_0.rlib --extern jobserver=output-1.54.0/cargo-build/libjobserver-0_1_22.rlib --extern lazycell=output-1.54.0/cargo-build/liblazycell-1_3_0.rlib --extern libc=output-1.54.0/cargo-build/liblibc-0_2_95_H21.rlib --extern log=output-1.54.0/cargo-build/liblog-0_4_14_H20000.rlib --extern libgit2_sys=output-1.54.0/cargo-build/liblibgit2_sys-0_12_18_H1f.rlib --extern memchr=output-1.54.0/cargo-build/libmemchr-2_4_0_H6.rlib --extern num_cpus=output-1.54.0/cargo-build/libnum_cpus-1_13_0.rlib --extern opener=output-1.54.0/cargo-build/libopener-0_4_1.rlib --extern percent_encoding=output-1.54.0/cargo-build/libpercent_encoding-2_1_0.rlib --extern rustfix=output-1.54.0/cargo-build/librustfix-0_5_1.rlib --extern semver=output-1.54.0/cargo-build/libsemver-1_0_3_H3.rlib --extern serde=output-1.54.0/cargo-build/libserde-1_0_126_H1a.rlib --extern serde_ignored=output-1.54.0/cargo-build/libserde_ignored-0_1_2.rlib --extern serde_json=output-1.54.0/cargo-build/libserde_json-1_0_64_He0.rlib --extern shell_escape=output-1.54.0/cargo-build/libshell_escape-0_1_5.rlib --extern strip_ansi_escapes=output-1.54.0/cargo-build/libstrip_ansi_escapes-0_1_0.rlib --extern tar=output-1.54.0/cargo-build/libtar-0_4_35.rlib --extern tempfile=output-1.54.0/cargo-build/libtempfile-3_2_0.rlib --extern termcolor=output-1.54.0/cargo-build/libtermcolor-1_1_2.rlib --extern toml=output-1.54.0/cargo-build/libtoml-0_5_7.rlib --extern unicode_xid=output-1.54.0/cargo-build/libunicode_xid-0_2_2.rlib --extern url=output-1.54.0/cargo-build/liburl-2_2_2_H1.rlib --extern walkdir=output-1.54.0/cargo-build/libwalkdir-2_3_2.rlib --extern clap=output-1.54.0/cargo-build/libclap-2_33_3_H260b.rlib --extern unicode_width=output-1.54.0/cargo-build/libunicode_width-0_1_8.rlib --extern im_rc=output-1.54.0/cargo-build/libim_rc-15_0_0.rlib --extern itertools=output-1.54.0/cargo-build/libitertools-0_10_0_H3.rlib --extern rustc_workspace_hack=output-1.54.0/cargo-build/librustc_workspace_hack-1_0_0.rlib --extern rand=output-1.54.0/cargo-build/librand-0_8_3_H1847.rlib
BUILD FAILED
make: *** [output-1.54.0/cargo] Error 1
```
if I run a grep against output folder by this symbol I have:
```
catap@Kirills-mini-m1 mrustc % grep -irn _ZRG2cE14curl0_4_36_H22B_09INIT_CTOR0g output-1.54.0                                      
Binary file output-1.54.0/cargo-build/libcurl-0_4_36_H22.rlib.o matches
output-1.54.0/cargo-build/libcurl-0_4_36_H22.rlib.c:16307:__attribute__((section("__DATA,__mod_init_func"))) union u_static_ZRG2cE14curl0_4_36_H22B_09INIT_CTOR0g{ t_ZRTFe1C0T0 val; uintptr_t raw[1]; } ZRG2cE14curl0_4_36_H22B_09INIT_CTOR0g;      // static ::"curl-0_4_36_H22"::#0::INIT_CTOR : extern "C" fn() -> ()
output-1.54.0/cargo-build/libcurl-0_4_36_H22.rlib.c:16366:union u_static_ZRG2cE14curl0_4_36_H22B_09INIT_CTOR0g ZRG2cE14curl0_4_36_H22B_09INIT_CTOR0g = { .raw = {(uintptr_t)&ZRG2cE14curl0_4_36_H22B_010init_inner0g} };     // static ::"curl-0_4_36_H22"::#0::INIT_CTOR : extern "C" fn() -> () = 0010000000000000{@0x0+8 = &::"curl-0_4_36_H22"::#0::init_inner}
catap@Kirills-mini-m1 mrustc % 
```
the only binary which contains such symbol is `arm64` and nm says:
```
catap@Kirills-mini-m1 mrustc % nm output-1.54.0/cargo-build/libcurl-0_4_36_H22.rlib.o | grep _ZRG2cE14curl0_4_36_H22B_09INIT_CTOR0g
0000000000018850 S _ZRG2cE14curl0_4_36_H22B_09INIT_CTOR0g
catap@Kirills-mini-m1 mrustc % lipo -info output-1.54.0/cargo-build/libcurl-0_4_36_H22.rlib.o
Non-fat file: output-1.54.0/cargo-build/libcurl-0_4_36_H22.rlib.o is architecture: arm64
catap@Kirills-mini-m1 mrustc %
```

--- a/rustc-1.54.0-src.patch
+++ b/rustc-1.54.0-src.patch
@@ -115,3 +115,137 @@
 +#[cfg(any(miri, rust_compiler = "mrustc", not(all(feature = "simd", any(target_arch = "x86_64")))))]
  use self::generic as arch;
 
+#
+# Backport which is required to support arm64 on macOS 12
+# See: https://github.com/alexcrichton/curl-rust/commit/0aea09c428b9bc2bcf46da0fc33959fe3f03c74a
+#
+diff --git vendor/curl/src/lib.rs vendor/curl/src/lib.rs
+index 9f2e50ea9b..50eaba742b 100644
+--- vendor/curl/src/lib.rs
++++ vendor/curl/src/lib.rs
+@@ -82,6 +82,9 @@ pub mod easy;
+ pub mod multi;
+ mod panic;
+ 
++#[cfg(test)]
++static INITIALIZED: std::sync::atomic::AtomicBool = std::sync::atomic::AtomicBool::new(false);
++
+ /// Initializes the underlying libcurl library.
+ ///
+ /// The underlying libcurl library must be initialized before use, and must be
+@@ -102,46 +105,62 @@ pub fn init() {
+     /// Used to prevent concurrent or duplicate initialization.
+     static INIT: Once = Once::new();
+ 
+-    /// An exported constructor function. On supported platforms, this will be
+-    /// invoked automatically before the program's `main` is called.
+-    #[cfg_attr(
+-        any(target_os = "linux", target_os = "freebsd", target_os = "android"),
+-        link_section = ".init_array"
+-    )]
+-    #[cfg_attr(target_os = "macos", link_section = "__DATA,__mod_init_func")]
+-    #[cfg_attr(target_os = "windows", link_section = ".CRT$XCU")]
+-    static INIT_CTOR: extern "C" fn() = init_inner;
++    INIT.call_once(|| {
++        #[cfg(need_openssl_init)]
++        openssl_probe::init_ssl_cert_env_vars();
++        #[cfg(need_openssl_init)]
++        openssl_sys::init();
++
++        unsafe {
++            assert_eq!(curl_sys::curl_global_init(curl_sys::CURL_GLOBAL_ALL), 0);
++        }
++
++        #[cfg(test)]
++        {
++            INITIALIZED.store(true, std::sync::atomic::Ordering::SeqCst);
++        }
++
++        // Note that we explicitly don't schedule a call to
++        // `curl_global_cleanup`. The documentation for that function says
++        //
++        // > You must not call it when any other thread in the program (i.e. a
++        // > thread sharing the same memory) is running. This doesn't just mean
++        // > no other thread that is using libcurl.
++        //
++        // We can't ever be sure of that, so unfortunately we can't call the
++        // function.
++    });
++}
+ 
++/// An exported constructor function. On supported platforms, this will be
++/// invoked automatically before the program's `main` is called. This is done
++/// for the convenience of library users since otherwise the thread-safety rules
++/// around initialization can be difficult to fulfill.
++///
++/// This is a hidden public item to ensure the symbol isn't optimized away by a
++/// rustc/LLVM bug: https://github.com/rust-lang/rust/issues/47384. As long as
++/// any item in this module is used by the final binary (which `init` will be)
++/// then this symbol should be preserved.
++#[used]
++#[doc(hidden)]
++#[cfg_attr(
++    any(target_os = "linux", target_os = "freebsd", target_os = "android"),
++    link_section = ".init_array"
++)]
++#[cfg_attr(target_os = "macos", link_section = "__DATA,__mod_init_func")]
++#[cfg_attr(target_os = "windows", link_section = ".CRT$XCU")]
++pub static INIT_CTOR: extern "C" fn() = {
+     /// This is the body of our constructor function.
+     #[cfg_attr(
+         any(target_os = "linux", target_os = "android"),
+         link_section = ".text.startup"
+     )]
+-    extern "C" fn init_inner() {
+-        INIT.call_once(|| {
+-            #[cfg(need_openssl_init)]
+-            openssl_sys::init();
+-
+-            unsafe {
+-                assert_eq!(curl_sys::curl_global_init(curl_sys::CURL_GLOBAL_ALL), 0);
+-            }
+-
+-            // Note that we explicitly don't schedule a call to
+-            // `curl_global_cleanup`. The documentation for that function says
+-            //
+-            // > You must not call it when any other thread in the program (i.e.
+-            // > a thread sharing the same memory) is running. This doesn't just
+-            // > mean no other thread that is using libcurl.
+-            //
+-            // We can't ever be sure of that, so unfortunately we can't call the
+-            // function.
+-        });
++    extern "C" fn init_ctor() {
++        init();
+     }
+ 
+-    // We invoke our init function through our static to ensure the symbol isn't
+-    // optimized away by a bug: https://github.com/rust-lang/rust/issues/47384
+-    INIT_CTOR();
+-}
++    init_ctor
++};
+ 
+ unsafe fn opt_str<'a>(ptr: *const libc::c_char) -> Option<&'a str> {
+     if ptr.is_null() {
+@@ -158,3 +177,20 @@ fn cvt(r: curl_sys::CURLcode) -> Result<(), Error> {
+         Err(Error::new(r))
+     }
+ }
++
++#[cfg(test)]
++mod tests {
++    use super::*;
++
++    #[test]
++    #[cfg(any(
++        target_os = "linux",
++        target_os = "macos",
++        target_os = "windows",
++        target_os = "freebsd",
++        target_os = "android"
++    ))]
++    fn is_initialized_before_main() {
++        assert!(INITIALIZED.load(std::sync::atomic::Ordering::SeqCst));
++    }
++}

